File: packages/java/src/main/java/com/causet/sdk/Emitter.java
package com.causet.sdk;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

/**
 * Minimal thread-safe pub/sub bus, mirroring the Emitter used by the
 * TypeScript, Python, and Laravel SDKs (on/emit).
 */
public class Emitter {
    private final Map<String, CopyOnWriteArrayList<Consumer<Object>>> handlers = new ConcurrentHashMap<>();
    private final CopyOnWriteArrayList<BiConsumer<String, Object>> wildcards = new CopyOnWriteArrayList<>();

    /** Registers handler for eventType. Returns a Runnable that unsubscribes it. */
    public Runnable on(String eventType, Consumer<Object> handler) {
        List<Consumer<Object>> list = handlers.computeIfAbsent(eventType, k -> new CopyOnWriteArrayList<>());
        list.add(handler);
        return () -> list.remove(handler);
    }

    /** Registers a wildcard handler invoked for every emitted event. */
    public Runnable onAny(BiConsumer<String, Object> handler) {
        wildcards.add(handler);
        return () -> wildcards.remove(handler);
    }

    /** Invokes every handler registered for eventType, then every wildcard handler. */
    public void emit(String eventType, Object data) {
        for (Consumer<Object> h : handlers.getOrDefault(eventType, EMPTY)) {
            try {
                h.accept(data);
            } catch (RuntimeException ignored) {
                /* a handler error must not break the caller (e.g. stream read loop) */
            }
        }
        for (BiConsumer<String, Object> h : wildcards) {
            try {
                h.accept(eventType, data);
            } catch (RuntimeException ignored) {
                /* ignored */
            }
        }
    }

    private static final CopyOnWriteArrayList<Consumer<Object>> EMPTY = new CopyOnWriteArrayList<>();
}

File: packages/laravel/src/Support/Emitter.php
<?php

namespace Causet\Laravel\Support;

class Emitter
{
    /** @var array<string, array<int, callable>> */
    private array $handlers = [];

    /** @var array<int, callable> */
    private array $wildcard = [];

    public function on(string $eventType, callable $handler): callable
    {
        if ($eventType === '*') {
            $this->wildcard[] = $handler;

            return function () use ($handler): void {
                $this->wildcard = array_values(array_filter(
                    $this->wildcard,
                    fn (callable $h): bool => $h !== $handler
                ));
            };
        }

        $this->handlers[$eventType] ??= [];
        $this->handlers[$eventType][] = $handler;

        return function () use ($eventType, $handler): void {
            if (! isset($this->handlers[$eventType])) {
                return;
            }
            $this->handlers[$eventType] = array_values(array_filter(
                $this->handlers[$eventType],
                fn (callable $h): bool => $h !== $handler
            ));
        };
    }

    public function emit(string $eventType, mixed $data = null): void
    {
        foreach ($this->handlers[$eventType] ?? [] as $handler) {
            try {
                $handler($data);
            } catch (\Throwable) {
                // handler errors are ignored
            }
        }

        foreach ($this->wildcard as $handler) {
            try {
                $handler($eventType, $data);
            } catch (\Throwable) {
                // handler errors are ignored
            }
        }
    }
}
